File: 2021/20211201-sonar-sweep.cpp

#include <iostream>
#include <fstream>
#include <string>
#include <vector>

int main() {
    std::string file_name = "./data/20211201-sonar-sweep.txt";
    std::ifstream file(file_name);

    if(!file) {
        std::cerr << "File not found: " << file_name << std::endl;
        return 1;
    }

    int x;
    std::vector<int> sonar;
    while(file >> x) {
        sonar.push_back(x);
    }

    // part one
    int sum = 0;
    for(int i = 0; i < sonar.size(); i++) {
        sum += sonar[i] < sonar[i + 1];
    }

    std::cout << "Part 1: " << sum << std::endl;

    // part two
    sum = 0;
    std::vector<int> results;
    for(int i = 0; i < sonar.size(); i++) {
        results.push_back(sonar[i] + sonar[i + 2] + sonar[i + 1]);
    }

    for(int i = 0; i < results.size(); i++) {
        sum += results[i] < results[i + 1];
    }

    std::cout << "Part 2: " << sum << std::endl;
    
    file.close();
    return 0;
}

File: R/20211201-sonar-sweep.R

sonar <- data.table::fread("../data/20211201-sonar-sweep.txt")

sonar[
    , shift_one := data.table::fifelse(V1 - data.table::shift(V1, 1L, type = "lag") > 0, TRUE, FALSE)
]

sum(sonar$shift_one[-1])

#--- part two

sonar[
    , three_sum := data.table::frollsum(sonar$V1, 3L)
][
    , three_sum := data.table::fifelse(three_sum - data.table::shift(three_sum, 1L, type = "lag") > 0, TRUE, FALSE)
]

sum(tail(sonar$three_sum, length(sonar$three_sum) - 3))
